eval/topdown.go

```diff
@@ -7,42 +7,34 @@ package eval
 import "github.com/open-policy-agent/opa/opalog"
 import "fmt"
 
-var _ = fmt.Println
-
-// Bindings is the structure used to ground expressions.
-type Bindings map[opalog.Var]opalog.Value
-
 // TopDownContext contains the state of the evaluation process.
 //
 // TODO(tsandall): profile perf/memory usage with current approach;
 // the current approach copies the TopDownContext structure for each
 // step and binding. This avoids the need to undo steps and bindings
 // each time the proof fails but this may be too expensive.
 type TopDownContext struct {
-	Rule     *opalog.Rule
-	Bindings Bindings
+	Query    opalog.Body
+	Bindings *hashMap
 	Index    int
 	Previous *TopDownContext
 	Store    Storage
+	Tracer   Tracer
 }
 
-// Current returns the current expression to evaluate.
-func (ctx *TopDownContext) Current() *opalog.Expr {
-	return ctx.Rule.Body[ctx.Index]
-}
-
-// Step returns a new context to evaluate the next expression.
-func (ctx *TopDownContext) Step() *TopDownContext {
-	return &TopDownContext{
-		Rule:     ctx.Rule,
-		Bindings: ctx.Bindings,
-		Index:    ctx.Index + 1,
-		Previous: ctx.Previous,
-		Store:    ctx.Store,
-	}
+// BindRef returns a new TopDownContext with bindings that map the reference to the value.
+func (ctx *TopDownContext) BindRef(ref opalog.Ref, value opalog.Value) *TopDownContext {
+	cpy := *ctx
+	cpy.Bindings = newHashMap()
+	ctx.Bindings.Iter(func(k opalog.Value, v opalog.Value) bool {
```

Comment:
Is this just copying ctx.Bindings into cpy.Bindings?  Maybe worth having just a 'copy' method in HashMap.